backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala

```diff
@@ -917,43 +917,106 @@ trait StandardAsyncExecutionActor
+  def retryElseFail(backendExecutionStatus: Future[ExecutionHandle],
                     retryWithMoreMemory: Boolean = false): Future[ExecutionHandle] = {
 
-    val retryable = previousFailedRetries < maxRetries
-
     backendExecutionStatus flatMap {
-      case failed: FailedNonRetryableExecutionHandle if retryable =>
-        incrementFailedRetryCount map { _ =>
-          val currentMemoryMultiplier = jobDescriptor.key.memoryMultiplier
-          (retryWithMoreMemory, memoryRetryFactor) match {
-            case (true, Some(multiplier)) => refineV[GreaterEqualOne](currentMemoryMultiplier.value * multiplier.value) match {
-              case Left(_) => FailedNonRetryableExecutionHandle(MemoryMultiplierNotPositive(jobDescriptor.key.tag, None, currentMemoryMultiplier, multiplier), failed.returnCode)
-              case Right(newMultiplier) => FailedRetryableExecutionHandle(failed.throwable, failed.returnCode, newMultiplier)
+      case failedRetryableOrNonRetryable: FailedExecutionHandle =>
+
+        val kvsFromPreviousAttempt = jobDescriptor.prefetchedKvStoreEntries.collect {
+          case (key: String, kvPair: KvPair) => key -> kvPair
+        }
+        val kvsForNextAttempt = failedRetryableOrNonRetryable.kvPairsToSave match {
+          case Some(kvPairs) => kvPairs.map {
+            case kvPair@KvPair(ScopedKey(_, _, key), _) => key -> kvPair
+          }.toMap
+          case None => Map.empty[String, KvPair]
+        }
+
+        val maxRetriesNotReachedYet = previousFailedRetries < maxRetries
+        failedRetryableOrNonRetryable match {
+          case failed: FailedNonRetryableExecutionHandle if maxRetriesNotReachedYet =>
+            val currentMemoryMultiplier = jobDescriptor.key.memoryMultiplier
+            (retryWithMoreMemory, memoryRetryFactor) match {
+              case (true, Some(multiplier)) => refineV[GreaterEqualOne](currentMemoryMultiplier.value * multiplier.value) match {
+                case Left(_) => Future.successful(
+                  FailedNonRetryableExecutionHandle(
+                    MemoryMultiplierNotPositive(jobDescriptor.key.tag, None, currentMemoryMultiplier, multiplier),
+                    failed.returnCode,
+                    None
+                  )
+                )
+                case Right(newMultiplier) => saveAttrsAndRetry(failed, kvsFromPreviousAttempt, kvsForNextAttempt, incFailedCount = true, Option(newMultiplier))
+              }
+              case (_, _) => saveAttrsAndRetry(failed, kvsFromPreviousAttempt, kvsForNextAttempt, incFailedCount = true)
             }
-            case (_, _) => FailedRetryableExecutionHandle(failed.throwable, failed.returnCode, currentMemoryMultiplier)
-          }
+          case failedNonRetryable: FailedNonRetryableExecutionHandle => Future.successful(failedNonRetryable)
+          case failedRetryable: FailedRetryableExecutionHandle => saveAttrsAndRetry(failedRetryable, kvsForNextAttempt, kvsForNextAttempt, incFailedCount = false)
```

Comment:
This is because if we get `FailedRetryableExecutionHandle` here, that means that job is retryable for some reason which is known and checked by backend, but which doesn't affect failed retries counter yet.
For example, we end up here, if preemption occurred, but the max number of preemption retries has not been exceeded yet (this is checked in the particular backend actor implementation, `PipelinesApiAsyncBackendJobExecutionActor`). In this case there will be already incremented preemption counter inside the `kvsForNextAttempt` parameter. 
But if number of preemptions has already reached the configured maximum, then we'll have `FailedNonRetryableExecutionHandle` here and we'll start checking the failed retries counter.

Reply:
A doc describing retry logic will be created under a separate ticket: BA-6139